src/vs/platform/terminal/electron-main/localPtyMainService.ts

```diff
@@ -86,15 +99,226 @@ export class LocalPtyMainService extends Disposable implements ICommonLocalPtySe
 		return this._throwIfNoPty(id).getCwd();
 	}
 
-	async getLatency(id: number): Promise<number> {
-		return this._throwIfNoPty(id).getLatency();
-	}
-
-	private _throwIfNoPty(id: number): ITerminalChildProcess {
+	private _throwIfNoPty(id: number): PersistentTerminalProcess {
 		const pty = this._localPtys.get(id);
 		if (!pty) {
 			throw new Error(`Could not find pty with id "${id}"`);
 		}
 		return pty;
 	}
 }
+
+export class PersistentTerminalProcess extends Disposable {
+
+	private readonly _events: Emitter<IRemoteTerminalProcessEvent>;
+	public readonly events: Event<IRemoteTerminalProcessEvent>;
+
+	private readonly _bufferer: TerminalDataBufferer;
+
+	private readonly _pendingCommands = new Map<number, { resolve: (data: any) => void; reject: (err: any) => void; }>();
+
+	private readonly _recorder: TerminalRecorder;
+	private _seenFirstListener: boolean;
+
+	private _orphanQuestionBarrier: AutoOpenBarrier | null;
+	private _orphanQuestionReplyTime: number;
+	private _orphanRequestQueue = new Queue<boolean>();
+	private _disconnectRunner1: RunOnceScheduler;
+	private _disconnectRunner2: RunOnceScheduler;
+
+	private _title = '';
+	private _pid = -1;
+
+	public get pid(): number {
+		return this._pid;
+	}
+
+	public get title(): string {
+		return this._title;
+	}
+
+	constructor(
+		private readonly _id: number,
+		private readonly _terminalProcess: TerminalProcess,
+		public readonly workspaceId: string,
+		public readonly workspaceName: string,
+		public readonly shouldPersistTerminal: boolean,
+		cols: number, rows: number,
+		ipcHandlePath: string,
+		private readonly _logService: ILogService,
+		private readonly _onExit: () => void,
+	) {
+		super();
+
+		this._recorder = new TerminalRecorder(cols, rows);
+		this._seenFirstListener = false;
+
+		this._orphanQuestionBarrier = null;
+		this._orphanQuestionReplyTime = 0;
+		this._disconnectRunner1 = this._register(new RunOnceScheduler(() => {
+			this._logService.info(`The reconnection grace time of ${printTime(ProtocolConstants.ReconnectionGraceTime)} has expired, so the terminal process with pid ${this._pid} will be shutdown.`);
+			this.shutdown(true);
+		}, ProtocolConstants.ReconnectionGraceTime));
+		this._disconnectRunner2 = this._register(new RunOnceScheduler(() => {
+			this._logService.info(`The short reconnection grace time of ${printTime(ProtocolConstants.ReconnectionShortGraceTime)} has expired, so the terminal process with pid ${this._pid} will be shutdown.`);
+			this.shutdown(true);
+		}, ProtocolConstants.ReconnectionShortGraceTime));
```

Comment:
These times should come in through the ctor as they'll be much more aggressive for the local case, we basically only want to make sure processes are restored across reloads and make sure they're killed asap if a reload did not occur.